torch_xla/csrc/nll_loss.cpp

```diff
@@ -41,49 +41,66 @@ xla::XlaOp LabelsToOneHot(xla::XlaBuilder* builder, xla::int64 depth, int axis,
                      xla::Broadcast(off_value, output_dimensions));
 }
 
+// Count the number of labels which aren't equal to "ignore_index".
+xla::XlaOp ValidLabelsCount(const xla::XlaOp& labels, int ignore_index) {
+  xla::XlaBuilder* builder = labels.builder();
+  xla::Shape labels_shape = XlaHelpers::ShapeOfXlaOp(labels);
+  xla::XlaOp valid_bitmap =
+      xla::Ne(labels, XlaHelpers::ScalarValue<xla::int64>(
+                          ignore_index, labels_shape.element_type(), builder));
+  valid_bitmap = xla::ConvertElementType(valid_bitmap, xla::PrimitiveType::S32);
+  xla::XlaOp zero =
+      XlaHelpers::ScalarValue<xla::int32>(0, xla::PrimitiveType::S32, builder);
+  xla::XlaComputation add_func =
+      XlaHelpers::CreateAddComputation(xla::PrimitiveType::S32);
+  return xla::ReduceAll(valid_bitmap, zero, add_func);
+}
+
 }  // namespace
 
 // Builds the NLLLoss for log-probabilities "logits" and class indices "labels".
-xla::XlaOp BuildNllLoss(const xla::XlaOp& logits, const xla::XlaOp& labels) {
+xla::XlaOp BuildNllLoss(const xla::XlaOp& logits, const xla::XlaOp& labels,
+                        int ignore_index) {
   xla::XlaBuilder* builder = logits.builder();
   xla::Shape logits_shape = XlaHelpers::ShapeOfXlaOp(logits);
+  xla::PrimitiveType logits_element_type = logits_shape.element_type();
   xla::XlaOp zero =
-      XlaHelpers::ScalarValue<float>(0, logits_shape.element_type(), builder);
+      XlaHelpers::ScalarValue<float>(0, logits_element_type, builder);
   xla::XlaOp one_hot_labels = LabelsToOneHot(
       /*builder=*/builder,
       /*depth=*/logits_shape.dimensions(1),
       /*axis=*/1,
       /*indices=*/labels,
       /*on_value=*/
-      XlaHelpers::ScalarValue<float>(1, logits_shape.element_type(), builder),
+      XlaHelpers::ScalarValue<float>(1, logits_element_type, builder),
       /*off_value=*/zero);
   // Compute sum(-one_hot_labels * logits) / batch.
   xla::XlaOp mul = xla::Mul(xla::Neg(one_hot_labels), logits);
+  xla::XlaOp batch = xla::ConvertElementType(
+      ValidLabelsCount(labels, ignore_index), logits_element_type);
   xla::XlaComputation add_func =
-      XlaHelpers::CreateAddComputation(logits_shape.element_type());
-  xla::XlaOp batch = XlaHelpers::ScalarValue<float>(
-      logits_shape.dimensions(0), logits_shape.element_type(), builder);
+      XlaHelpers::CreateAddComputation(logits_element_type);
   return xla::ReduceAll(mul, zero, add_func) / batch;
```

Comment:
I may be mistaken, but I think the values with ignore_index should not be reduced.
IOW, doing the proper "batch" division is only one part of the job.
Also, how is our one-hot WRT negative class indices?
Can you try to test with an ignore_index within the range of the class indices (say, ignore_index == 5 in a 10 class setup)?

Reply:
I'll double check this. If it's outside the range, it'll work out ok since one-hot will be all zeros for that batch, but if it's inside the range you might have be right.

Reply:
Done, added a path for the case when `ignore_index` is a valid class index.